core/src/main/scala/org/http4s/internal/parsing/Rfc8941.scala

```diff
@@ -0,0 +1,169 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.internal.parsing
+
+import cats.parse.Parser
+import cats.parse.Parser0
+import cats.parse.Rfc5234
+
+/** Parsers for the rules defined in RFC8941
+  *
+  * @see [[https://datatracker.ietf.org/doc/html/rfc8941]]
+  */
+private[http4s] object Rfc8941 {
+
+  /* sf-integer = ["-"] 1*15DIGIT
+   */
+  val sfInteger: Parser[Long] = {
+    import Parser.char, Rfc5234.digit
+    val pos = digit.rep(1, 15).string
+    val neg = (char('-') *> pos).map(s => s"-$s")
+    (pos | neg).map(_.toLong)
+  }
+
+  /* sf-decimal = ["-"] 1*12DIGIT "." 1*3DIGIT
+   */
+  val sfDecimal: Parser[BigDecimal] = {
+    import Parser.char, Rfc5234.digit
+    val num = digit.rep(1, 12).string
+    val dec = digit.rep(1, 3).string
+    val pos = (num ~ (char('.') *> dec)).map(t => s"${t._1}.${t._2}")
+    val neg = (char('-') *> pos).map(s => s"-$s")
+    (pos | neg).map(BigDecimal(_))
+  }
+
+  /* sf-string = DQUOTE *chr DQUOTE
+   * chr       = unescaped / escaped
+   * unescaped = %x20-21 / %x23-5B / %x5D-7E
+   * escaped   = "\" ( DQUOTE / "\" )
+   */
+  val sfString: Parser[String] = {
+    import Parser.{charIn, string}, Rfc5234.dquote
+    val escaped = string("\\\\") | string("\\\"")
+    val unescaped =
+      charIn(0x20.toChar to 0x21.toChar)
+        .orElse(charIn(0x23.toChar to 0x5b.toChar))
+        .orElse(charIn(0x5d.toChar to 0x7e.toChar))
+        .string
+    val chr = unescaped | escaped
+    (dquote *> chr.rep0.string <* dquote).map(s => s""""$s"""")
+  }
+
+  /* sf-token = ( ALPHA / "*" ) *( tchar / ":" / "/" )
+   */
+  val sfToken: Parser[String] = {
+    import Parser.{char, charIn}, Rfc5234.alpha, Rfc7230.tchar
+    val head = (alpha | char('*')).string
+    val tail = (tchar | charIn(":/".toList)).rep0.string
+    (head ~ tail).map(t => t._1 + t._2)
+  }
+
+  /* sf-binary = ":" *(base64) ":"
+   * base64    = ALPHA / DIGIT / "+" / "/" / "="
+   */
+  val sfBinary: Parser[String] = {
+    import Parser.{char, charIn}, Rfc5234.{alpha, digit}
+    val base64 = (alpha | digit | charIn("+/=".toList)).rep0.string
+    char(':') *> base64 <* char(':')
+  }
+
+  /* sf-boolean = "?" boolean
+   * boolean    = "0" / "1"
+   */
+  val sfBoolean: Parser[Boolean] = {
+    import Parser.string
+    string("?0").as(false) | string("?1").as(true)
+  }
+
+  /* bare-item = sf-integer / sf-decimal / sf-string / sf-token
+   *           / sf-binary / sf-boolean
+   */
+  def bareItem[A](
+      sfIntegerP: Parser[A],
+      sfDecimalP: Parser[A],
+      sfStringP: Parser[A],
+      sfTokenP: Parser[A],
+      sfBinaryP: Parser[A],
+      sfBooleanP: Parser[A],
```

Comment:
Does this need the type parameter and parser parameters? Can't they be hard-coded?

Reply:
This way we can keep the RFC8941 parsers separate from our data model and avoid cyclic dependencies between them.

Reply:
Done.